x/oracle/types/currency_pair.go

```diff
@@ -31,48 +34,88 @@ func (cp *CurrencyPair) ValidateBasic() error {
 	if strings.ToUpper(cp.Quote) != cp.Quote {
 		return fmt.Errorf("incorrectly formatted quote string, expected: %s got: %s", strings.ToUpper(cp.Quote), cp.Quote)
 	}
+
+	if cp.Decimals <= 0 {
+		return fmt.Errorf("decimals must be greater than 0")
+	}
+
 	return nil
 }
 
-// String returns a string representation of the CurrencyPair, in the following form "ETH/BTC".
+// String returns a string representation of the CurrencyPair, in the following form "ETH/BTC/8" where 8 is the number of decimals.
 func (cp CurrencyPair) String() string {
+	return fmt.Sprintf("%s/%s/%d", cp.Base, cp.Quote, cp.Decimals)
+}
+
+// ID returns a string representation of the CurrencyPair, in the following form "ETH/BTC" where 8 is the number of decimals.
+// This function differs from String() which also includes the decimal information.
+func (cp *CurrencyPair) ID() string {
 	return fmt.Sprintf("%s/%s", cp.Base, cp.Quote)
 }
 
+// CurrencyPairStringToID takes a CurrencyPair string and returns the ID representation of it.
+func CurrencyPairStringToID(cpStr string) (string, error) {
+	cp, err := CurrencyPairFromString(cpStr)
+	if err != nil {
+		split := strings.Split(cpStr, "/")
+		if len(split) == 2 {
+			return cpStr, nil
+		}
+
+		return "", fmt.Errorf("invalid string provided: %w", err)
+	}
+
+	return cp.ID(), nil
+}
+
 // CurrencyPairString constructs and returns the string representation of a currency pair.
-func CurrencyPairString(base, quote string) string {
-	cp := NewCurrencyPair(base, quote)
+func CurrencyPairString(base, quote string, decimals int64) string {
+	cp := NewCurrencyPair(base, quote, decimals)
 	return cp.String()
 }
 
 func CurrencyPairFromString(s string) (CurrencyPair, error) {
 	split := strings.Split(s, "/")
-	if len(split) != 2 {
+	if len(split) != 3 {
 		return CurrencyPair{}, fmt.Errorf("incorrectly formatted CurrencyPair: %s", s)
 	}
+
+	decimals, err := strconv.ParseInt(split[2], 10, 64)
+	if err != nil {
+		return CurrencyPair{}, fmt.Errorf("incorrectly formatted CurrencyPair: %s", s)
+	}
+
 	cp := CurrencyPair{
-		Base:  strings.ToUpper(split[0]),
-		Quote: strings.ToUpper(split[1]),
+		Base:     strings.ToUpper(split[0]),
+		Quote:    strings.ToUpper(split[1]),
```

Comment:
upper casing here introduces weird consequence where the base and quote are not correctly formatted but will not return an error.